Added computing matrix condition numbers (linalg.cond)

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.cond` for NumPy compatibility.

Ref #42666.

IvanYashchuk committed:
Added linalg_cond

IvanYashchuk committed:
Updated the implementation.

Now for p=None and p=±2 condition number is computed with svd.
Added errorchecks.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk committed:
Updated non-invertible case implementation

IvanYashchuk committed:
Updated docs

IvanYashchuk committed:
Added entry to common_methods_invocations.py

IvanYashchuk committed:
Added overrides.py entry

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Implement a trick for replacing FLT_MAX or DBL_MAX to INFINITY

IvanYashchuk committed:
Added out= variant

IvanYashchuk committed:
Updated implementation and tests. Complex input is now supported.

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Trying to fix initialization of 'dim'

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk committed:
Trying to fix initialization of 'dim'

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+}
+
+Tensor& linalg_cond_out(Tensor& result, const Tensor& self, optional<Scalar> opt_ord) {
+  ScalarType real_dtype = toValueType(typeMetaToScalarType(self.dtype()));
```
Add a comment here, why `real_dtype` is used.
When SVD is used to compute the condition number the result is always real-valued, for other cases it can be complex-valued.

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -410,6 +410,103 @@ def run_test_case(input, p, dim, keepdim):
+                continue
+            run_test_case(a, ord)
+
+    # TODO: once "inverse_cuda" supports complex dtypes, they shall be added to above tests
```
https://github.com/pytorch/pytorch/pull/45034 got merged. This needs to be updated.

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    squareCheckInputs(self);
+    std::array<int64_t, 2> dim_arr = {-2, -1};
+    optional<IntArrayRef> dim = IntArrayRef(dim_arr);
+    // Ignore errors if not invertible, result is INFINITY in this case
```
Probably should add a note here that if in the future it is decided that inverse routine shouldn't throw an error by default (for performance reasons). Then the behaviour here could be changed that the inverse matrix contains NaNs and the norm of such matrix is also NaN, and apply `at::nan_to_num(result, INFINITY);` to get the result.

mruberry replied:
We'll keep throwing errors by default but add `_ex` variants that return the errors instead of throwing them.

IvanYashchuk commented:
@mruberry, this PR is ready for starting the review process.
It adds `torch.linalg.cond` a function that computes the condition number, that is given a matrix `A` compute `norm(A)*norm(inverse(A))`.
`out=` variant is implemented. Autograd and JIT tests pass as well.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+cond = _add_docstr(_linalg.linalg_cond, r"""
+linalg.norm(input, p=None, *, out=None) -> Tensor
+
+Returns the condition number of a matrix.
```
Does input actually have to be a matrix? Does it have to be square?

This first sentence should define "condition number" OR be immediately followed by a mathematical statement describing it that includes input and p.

The first sentence should also reference input and p. For example: "Computes the condition number :attr:\`\input\`, a matrix, using the norm defined by :attr:\`p\`. The condition number is defined as: "

IvanYashchuk replied:
Thank you for the suggestion. I've changed the text now.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+        =====  ============================
+        ord    norm for matrices
+        =====  ============================
+        None   2-norm, computed directly using the SVD
```
Is None != 2?

IvanYashchuk replied:
p=None is the same as p=2. I updated the description and it's now the same.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+This function supports real-valued, and only on CPU, complex-valued input.
+
+Args:
+    input (Tensor): the input tensor of size :math:`(*, m, n)` where `*` is zero or more
```
More accurately it's an input matrix of size (m, n) or a batch of matrices of size (l, m, n). If the star were actually zero the tensor would have no elements and would hit your runtime check above.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+
+Args:
+    input (Tensor): the input tensor of size :math:`(*, m, n)` where `*` is zero or more
+                    batch dimensions.
```
It's not multiple dimensions, it's just that the batch may contain multiple matrices.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+    input (Tensor): the input tensor of size :math:`(*, m, n)` where `*` is zero or more
+                    batch dimensions.
+
+    p (int, float, inf, -inf, 'fro', optional): The order of norm.
```
Is Frobenius an "order" of a norm? 

More accurately, the p argument describes which norm to use.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+                    batch dimensions.
+
+    p (int, float, inf, -inf, 'fro', optional): The order of norm.
+        inf refers to :attr:`float('inf')`, numpy's :attr:`inf` object, or any equivalent object.
```
I think we can skip this detail here.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,66 @@
+
+    p (int, float, inf, -inf, 'fro', optional): The order of norm.
+        inf refers to :attr:`float('inf')`, numpy's :attr:`inf` object, or any equivalent object.
+        The following norms can be used:
```
"can be used" -> "are supported"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+// Numerical or None norms
+Tensor linalg_cond(const Tensor& self, optional<Scalar> opt_ord) {
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  TORCH_CHECK(self.dim() >= 2, "Tensor of matrices must have at least 2 dimensions.");
```
"cond only supports matrices (dims=2) or batches of matrices (dims=3), but got a tensor with dims=X"

Would this work on a 4D tensor, too?

IvanYashchuk replied:
It works for ND, N>2, tensors. 4D tensors are tested to work.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    if (ord.toDouble() == -2.0) {
+      result = s_min / s_max;
+    }
+    else {
```
Egyptian-style with else on the same line as the closing brace

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    return result;
+  }
+  // ord == ±1 ord == ±inf
+  else {
```
I don't think you actually need the else branch here because the previous branch always returns

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    // a trick to convert FLT_MAX or DBL_MAX to INFINITY
+    // result might also contains infs already then inf + inf - inf = nan
+    // this will get replaced by inf
+    result = result + result - result;
```
I think some of these operations can be performed inplace to save memory. 

Can you elaborate on what's going on here? What happens if result is a float dtype and contains values in the latter half of float's dynamic range?

IvanYashchuk replied:
That was clearly incorrect.
I've changed it now to `at::where(result == DBL_MAX/FLT_MAX, INFINITY, result)`. That's needed for NumPy compatiblity.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  }
+  // ord == ±1 ord == ±inf
+  else {
+    squareCheckInputs(self);
```
So the matrices do have to be square?

IvanYashchuk replied:
Yes, for p = {±1, ±inf, 'fro', 'nuc'}, because `at::inverse` is used to compute the inverse matrix. `at::inverse` requires square matrices as input.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    } catch (const std::exception& e) {
+      if (strstr(e.what(), "singular")) {
+      result = at::empty_like(at::linalg_norm(self, ord, dim));
+      at::fill_(result, INFINITY);
```
What happens if the input was a batch of matrices?

mruberry replied:
You mean we should raise an error if ANY of the matrices is not invertible? That seems correct. 

This PR should then add a test validating this discrepancy with NumPy and the behavior should be documented.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+      self_inverse = at::inverse(self);
+    } catch (const std::exception& e) {
+      if (strstr(e.what(), "singular")) {
+      result = at::empty_like(at::linalg_norm(self, ord, dim));
```
This is interesting. Why do you need to call at::linalg_norm here? That seems inefficient.

IvanYashchuk replied:
I've changed it now.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    optional<IntArrayRef> dim = IntArrayRef(dim_arr);
+    // Ignore errors if not invertible, result is INFINITY in this case
+    try {
+      self_inverse = at::inverse(self);
```
This will cause cross-device data movement when performing the error check, right? We should document this behavior.

IvanYashchuk replied:
That's right, currently `at::inverse` always checks for error and this causes device-host memory synchronization. I added the comment on this.

mruberry replied:
Thanks @IvanYashchuk. cc @heitorschueroff

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  TORCH_CHECK(self.dim() >= 2, "Tensor of matrices must have at least 2 dimensions.");
+
+  Tensor self_inverse, result;
```
The declaration of self_inverse can go into the second block after the {-2, 2}-norm case

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+// Frobenius or nuclear norms
+Tensor linalg_cond(const Tensor& self, std::string ord) {
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  squareCheckInputs(self);
```
No dims check? What is squareCheckInputs actually enforcing here?

IvanYashchuk replied:
`squareCheckInputs` includes dims check and enforces that the last two dimensions have the same size.
I replaced now`squareCheckInputs` with the same checks but with a slightly modified error message.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,110 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  squareCheckInputs(self);
+
+  Tensor self_inverse, result;
```
Similar comments/questions here as in the above impl.

mruberry commented on test/test_linalg.py:
```diff
@@ -410,6 +410,103 @@ def run_test_case(input, p, dim, keepdim):
+            for ord in norm_types:
+                # frobenius norm not supported for complex tensors
+                if dtype.is_complex and ord == 'fro':
+                    continue
```
Instead of continuing verify it throws an error

mruberry commented on test/test_linalg.py:
```diff
@@ -410,6 +410,103 @@ def run_test_case(input, p, dim, keepdim):
+            self.assertEqual(ans, result)
+
+        norm_types = [1, -1, 2, -2, inf, -inf, 'fro', 'nuc', None]
+        input_sizes = [(3, 3), (2, 3, 3, 3)]
```
Vary the last two dimensions between the examples and, if this supports non-square matrices, create a non-square case, too

IvanYashchuk replied:
Non-square matrices are supported for {None, 2, -2} norms. I added it now to the tests.

mruberry commented on test/test_linalg.py:
```diff
@@ -410,6 +410,103 @@ def run_test_case(input, p, dim, keepdim):
+    @skipCUDAIfNoMagma
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    @dtypesIfCUDA(torch.float32, torch.float64)
+    @precisionOverride({torch.float32: 1e-3})
```
Is this because NumPy computes the condition number in double?

IvanYashchuk replied:
Yes, NumPy computes SVD and the inverse matrix in double precision.

mruberry reviewed (commented):
Hey @IvanYashchuk! Looking forward to having a NumPy-compatible condition number calculation in PyTorch. 

I made some comments and asked a few questions. Docs need a few refinements and I think the implementations can be made more efficient and consistent, too.

IvanYashchuk committed:
Small changes in the implementation structure

IvanYashchuk committed:
Use the explicit shape for the result instead of dummy norm computations

IvanYashchuk committed:
Fixed conversion FLT_MAX / DBL_MAX -> INFINITY

IvanYashchuk committed:
Added a warning for batched singular input; slightly refactored the tests

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Added non-square input test cases

IvanYashchuk committed:
Updated norms description

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk commented:
@mruberry, I've updated the implementation and documentation. I also added non-square input test cases.

And there is a question on singular batch input https://github.com/pytorch/pytorch/pull/45832#discussion_r520677707

mruberry commented:
Hey @IvanYashchuk, is this ready for another review or does the error-checking behavior need to be updated still?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk committed:
Raise error for batched input with at least one non-invertible matrix

IvanYashchuk commented:
@mruberry, I added a not-implemented-error for the batched input with at least one non-invertible matrix. It will be possible to implement the NumPy matching behavior once there is `at::inverse` that doesn't throw errors and calculates something for the whole input.

Now this is ready for another review.

mruberry commented:
> @mruberry, I added a not-implemented-error for the batched input with at least one non-invertible matrix. It will be possible to implement the NumPy matching behavior once there is `at::inverse` that doesn't throw errors and calculates something for the whole input.
> 
> Now this is ready for another review.

Hope you had a good weekend, @IvanYashchuk. I'll take a look now.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,73 @@
+This function supports real-valued, and only on CPU, complex-valued input.
+
+.. note:: For `p = {None, 2, -2}` :attr:`input` can be non-square. For other norm types :attr:`input` must be
+          a square matrix or a batch of square matrices. If :attr:`input` does not satisfy the requirement
```
"not satisfy the" -> "not satisfy this"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,73 @@
+          a square matrix or a batch of square matrices. If :attr:`input` does not satisfy the requirement
+          then a RuntimeError will be thrown.
+
+.. note:: For the batched input if at least one matrix in the batch is not invertible,
```
"For the batched input if at least one matrix in the batch is not invertible..." -> "If :attr:\`input\` is a non-invertible matrix or :attr:\`input\` is a batch of matrices and one or more of them is not invertible then a RuntimeError will be thrown.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,73 @@
+
+.. note:: For the batched input if at least one matrix in the batch is not invertible,
+          currently getting the result for all other (possibly) invertible matrices in the batch is not implemented.
+
```
A warning about the cross-device data movement needs to appear here, too. Like we did for Cholesky (https://pytorch.org/docs/master/linalg.html?highlight=cholesky#torch.linalg.cholesky). 

However I would suggest we simplify the warning message:

"When given inputs on a CUDA device, this function synchronizes that device with the CPU."

cc @heitorschueroff who will be reviewing these warnings holistically later.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,73 @@
+Examples::
+
+    >>> from torch import linalg as LA
+    >>> a = torch.tensor([[1., 0, -1], [0, 1, 0], [1, 0, 1]])
```
This is a good example but let's add a batched example, too

IvanYashchuk replied:
I added a batched example, and one more with a complex dtype. I didn't repeat all the norm types.
```python
    >>> a = torch.randn(3, 4, 4)
    >>> LA.cond(a)
    tensor([ 4.4739, 76.5234, 10.8409])

    >>> a = torch.randn(3, 4, 4, dtype=torch.complex64)
    >>> LA.cond(a)
    tensor([ 5.9175, 48.4590,  5.6443])
    >>> LA.cond(a, 1)
    >>> tensor([ 11.6734+0.j, 105.1037+0.j,  10.1978+0.j])
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,73 @@
+The condition number is defined as the matrix norm of
+:attr:`input` times the matrix norm of the inverse of :attr:`input`.
+
+This function supports real-valued, and only on CPU, complex-valued input.
```
"real-valued" -> "floating point", since I don't think this supports integer inputs?

mruberry replied:
Even more precisely, "float and double"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // This should change when at::inverse works with silent errors
+  if (self.dim() > 2) {
+    TORCH_CHECK(false,
+      "At least one matrix in the batch is not invertible, its condition number is infinity, "
```
Can we tell which matrix or which matrices aren't invertible? If so, we should include the first non-invertible matrix in this error message and the total number of non-invertible matrices.

If we can't tell, I suggest simplify this to:

"One or more matrices in the batch was not invertible!"

IvanYashchuk replied:
No, we can't tell how many matrices are non-invertible and it's not easy to extract the index of the first non-invertible one either.
I'll use your suggestion.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  auto result_shape = self.sizes().vec();
+  result_shape.pop_back();
+  result_shape.pop_back();  // result's shape is equal to self.shape[0:-2]
+  Tensor result = at::empty(result_shape, self.options());
```
Instead of empty + fill I think you can use full

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+        // Scalar.dtype() is always ScalarType::Double for isFloatingPoint() = true
+        // and at::where doesn't allow arguments with different dtype
+        // so let's use 0-dim tensor filled with inf
+        Tensor inf_tensor = at::empty({}, result.options());
```
Should be able to replace empty + fill with full

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    } else {
+      result = s_max / s_min;
+    }
+    // convert FLT_MAX or DBL_MAX to INFINITY for NumPy compatibility
```
Should we let ourselves be divergent from NumPy here?

This is going to be a significant perf hit. And what if the condition number actually is double max? 

Can you elaborate on what kind of issues we'd run into if we didn't do this conversion?

IvanYashchuk replied:
When the matrix is non-invertible (singular) it means it has zero singular values. Singular values are always non-negative, so the minimal possible value is zero. Now when we divide something with 0 we should get inf (by IEEE 754), but at the C++ level this gives FLT_MAX or DBL_MAX.
While in Python everything seems to be as expected.
```python
In [1]: import torch
In [2]: a = torch.eye(3)
In [3]: a[-1, -1] = 0  # make 'a' singular
In [4]: u, s, v = a.svd()
In [5]: s
Out[5]: tensor([1., 1., 0.])
In [6]: s.max() / s.min()
Out[6]: tensor(inf)
```

mruberry replied:
Wait... you're saying that our division operator isn't IEEE 754 compliant?

Here's our CPU division kernel as a reference: https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L95

And our CUDA division kernel:
https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/cuda/BinaryMulDivKernel.cu#L47

Can you elaborate on the behavior you're seeing on each device type?

IvanYashchuk replied:
Alright, I'm sorry for the confusion. All this FLT_MAX/DBL_MAX conversion to inf is redundant here. I will remove it now. Of course the division operator is compliant with the IEEE 754.

For history:
Initially, I had only one return statement in the end and I had to take care of NANs from possibly non-invertible input. I decided to use `at::nan_to_num` assuming it does what's in the name - converting NANs to the specified value, BUT additionally, this function also converts +inf to FLT_MAX/DBL_MAX, I should have read the documentation. And then I "fixed" it by `result=result+result-result` for +inf this of course gave nan that is then converted back to +inf 😅. I mistakenly thought that the problem is somewhere else that lead to the current code with these redundant calls to `at::where(result == dbl_max, inf, result)`.

mruberry replied:
Whew. Glad to hear it.

`nan_to_num`s behavior is confusing! Unfortunately our `nan_to_num` is consistent with NumPy's https://numpy.org/doc/stable/reference/generated/numpy.nan_to_num.html?highlight=nan_to_num#numpy.nan_to_num.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  TORCH_CHECK(self.dim() >= 2, "linalg_cond only supports matrices or batches of matrices, but got a tensor with ",
+    self.dim(), " dimensions.");
+  TORCH_CHECK(self.size(-1) == self.size(-2),
+              "linalg_cond only supports square matrices or batches of square matrices "
```
This warning should be a little more restrictive based on the documentation, since the None, 2, and -2 norms do allow non-square inputs. In this case I think it's OK to say that the frobenius and nuclear norms only support square matrices or batches of square matrices.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // since at::inverse is used in the implementation, self has to be a tensor consisting of square matrices
+  // the same check as squareCheckInputs(self) but with a slightly more informative error message
+  TORCH_CHECK(self.size(-1) == self.size(-2),
+              "linalg_cond only supports square matrices or batches of square matrices "
```
See comment below, too. It's only some norms that only supports square matrices.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  return result;
+}
+
+Tensor& linalg_cond_out(Tensor& result, const Tensor& self, std::string ord) {
```
These two functions that take strings are very similar to the functions that take integers. What would you think of reusing the same function for both by using something like an enum to represent the norm values? Then the string signatures would just translate their inputs to the enum values and call the common function.

IvanYashchuk replied:
I removed the duplicated code, now using `c10::variant<Scalar, std::string>` + `c10::visit`. Using enums in this case would require conversion functions `linalg_cond: scalar or string` -> `norm_ord_to_enum` -> helper function -> `enum_to_norm_ord` -> `linalg_norm`.

mruberry replied:
The approach you took looks great.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk committed:
Updated documentations

IvanYashchuk committed:
Use enums to dispatch on norm type

IvanYashchuk committed:
Add missing imports

IvanYashchuk committed:
Enums are redundant here, use only c10::variant for norm dispatch

IvanYashchuk committed:
Simplify error message for batched non-invertibel case

IvanYashchuk committed:
Use at::full instead of empty + fill_

IvanYashchuk committed:
Make error message more specific about the norm types

IvanYashchuk commented:
@mruberry, I think I've addressed almost all the issues now, thank you for the suggestions!
The only thing that is left is NumPy compatibility for 2-norms https://github.com/pytorch/pytorch/pull/45832#discussion_r525233096

Using `c10::variant<Scalar, std::string>` + `c10::visit` I removed the code duplication due to different types of `ord`.

IvanYashchuk committed:
Removed unneeded conversions to infinity

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

mruberry commented:
> @mruberry, I think I've addressed almost all the issues now, thank you for the suggestions!

You're welcome. It's always great to work with you.

> The only thing that is left is NumPy compatibility for 2-norms [#45832 (comment)](https://github.com/pytorch/pytorch/pull/45832#discussion_r525233096)

Can you elaborate on this? I thought we resolved the linked comment. What's the problem with p=2? 

> Using `c10::variant<Scalar, std::string>` + `c10::visit` I removed the code duplication due to different types of `ord`.

Awesome.

IvanYashchuk commented:
> Can you elaborate on this? I thought we resolved the linked comment. What's the problem with p=2?

It is resolved now.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  if (self.dim() > 2) {
+    TORCH_CHECK(false,
+      "One or more matrices in the batch was not invertible! "
+      "linalg_cond does not support yet this case.");
```
Let's remove this second line.

mruberry replied:
Follow-up question, if the intent is to turn inf for every value, should this be a warning and not a check?

Maybe something like...

```
TORCH_WARN("One or more matrices in the batch was not invertible! This will cause the entire result to be filled with infs.");
```

?

IvanYashchuk replied:
We've decided here https://github.com/pytorch/pytorch/pull/45832#discussion_r520748109 that since returning inf for every matrix is not correct we should raise an error. The batched case will be properly implemented in the future once it's possible to call `at::inverse` with silent errors (this should be possible after https://github.com/pytorch/pytorch/pull/48261 is ready, similarly to how it's now possible to call `at::solve_out_info` without error checks [here](https://github.com/pytorch/pytorch/pull/48456/files#diff-76df047bca5c2d7a540bdcc4023d68641c4f15638d83c5c874d833b03f9636b9R437-R439).

mruberry replied:
OK; I'm conviced.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // the result is always real-valued, for other cases it is complex-valued for the complex-valued input.
+  ScalarType real_dtype = toValueType(typeMetaToScalarType(self.dtype()));
+  Scalar ord = opt_ord.has_value() ? opt_ord.value() : 2;
+  auto expected_dtype = std::abs(ord.toDouble()) == 2.0 ? real_dtype : self.scalar_type();
```
Technically instead of casting ord to double unconditionally we should first check that it's not complex and not an integer that will overflow, but that seems onerous. 

I wonder if the Scalar type could/does do that for us automatically? It might catch some errors with the handling of complex scalars.

IvanYashchuk replied:
Conversion is checked here https://github.com/pytorch/pytorch/blob/master/c10/core/Scalar.h#L49-L60 + https://github.com/pytorch/pytorch/blob/master/c10/util/TypeCast.h#L160-L170

mruberry commented on test/test_linalg.py:
```diff
@@ -637,6 +637,121 @@ def run_test_case(input, p, dim, keepdim):
+        for input_size in input_sizes:
+            input = torch.randn(*input_size, dtype=dtype, device=device)
+            for ord in [2, -2, None]:
+                run_test_case(input, ord)
```
Might as well assert that the other norms fail on non-square inputs

mruberry commented on test/test_linalg.py:
```diff
@@ -637,6 +637,121 @@ def run_test_case(input, p, dim, keepdim):
+        a = a.repeat(batch_dim, 1, 1)
+        a[0, -1, -1] = 0  # now a[0] is singular
+        for ord in [1, -1, inf, -inf, 'fro', 'nuc']:
+            with self.assertRaisesRegex(RuntimeError, "linalg_cond does not support yet"):
```
Pending the outcome of the above discussion, this may change to a warning and the output (of all infs) might be testable.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,86 @@
 """)
 
+cond = _add_docstr(_linalg.linalg_cond, r"""
+linalg.norm(input, p=None, *, out=None) -> Tensor
```
In native_functions.yaml the signature has `ord` instead of `p` -- I think it just needs to be updated for consistency?

IvanYashchuk replied:
It also would be much better if `linalg.cond` and `linalg.norm` had the same argument name for describing the norm type, and `ord` and `p` aren't good names, since for example Frobenius norm is neither an order `ord` nor the power `p` of the norm.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,86 @@
+The condition number is defined as the matrix norm of
+:attr:`input` times the matrix norm of the inverse of :attr:`input`.
+
+This function supports ``float``, ``double``, and only on CPU, ``cfloat`` and ``cdouble`` dtype for :attr:`input`.
```
"dtype" -> "dtypes"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,86 @@
+          then a RuntimeError will be thrown.
+
+.. note:: If :attr:`input` is a non-invertible matrix then a tensor containing infinity will be returned.
+          If :attr:`input` is a batch of matrices and one or more of them is not invertible
```
This may change. Throwing a runtime error is also fine, I'm curious to hear your thoughts.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,86 @@
+        1      max(sum(abs(x), dim=0))
+        -1     min(sum(abs(x), dim=0))
+        2      2-norm (ratio of the largest singular value to the smallest singular value)
+        -2     ratio of the smallest singular value to the largest singular value
```
It seems like the line above should just be the phrase in the parentheses to be consistent with the presentation of the -2 norm and the other norms. That is,

```
2   ratio of the largest singular value to the smallest singular value
-2 ratio of the smallest singular value to the largest singular value
```

mruberry reviewed (approved):
Nice work, @IvanYashchuk. This whole PR is very "sharp," and I really like what you did with the documentation. I have one small design question and made a few other suggestions, but they're very minor. I'm curious to hear your thoughts.

I'm going to import this now, however, to see if the method_tests need to be removed.

mruberry commented:
As feared, the method_test updates are incredibly helpful for validating this PR but, unfortunately, the internal system can't handle them. 

The PR adding jit tests to OpInfos should land soon, however, and then we can just use them exclusively.

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+      "One or more matrices in the batch was not invertible! "
+      "linalg_cond does not support yet this case.");
+  }
+  auto result_shape = self.sizes().vec();
```
```suggestion
  auto result_shape = c10::IntArrayRef(self.sizes().begin(), self.sizes().end()-2);
```

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+
+// Numerical or None norms
+Tensor linalg_cond(const Tensor& self, optional<Scalar> opt_ord) {
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
```
It makes sense to factor out shape checks for both overloads. Also, numpy accepts empty batch size, so `numel==0` is a wrong check:
```
In [11]: a=np.arange(0).reshape(0,3,3).astype(np.float)                                                                                                                                                             

In [12]: a                                                                                                                                                                                                          
Out[12]: array([], shape=(0, 3, 3), dtype=float64)

In [13]: LA.cond(a)                                                                                                                                                                                                 
Out[13]: array([], dtype=float64)
```

IvanYashchuk replied:
Okay, I didn't check the NumPy's behavior for empty batch size, thank you!
NumPy doesn't work for empty matrices https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py#L1762-L1763.

mruberry replied:
It's likely that someone will ask we support empty matrices, too, and this PR can add that support as well (and test it and note in the tests its divergent from NumPy).

IvanYashchuk replied:
I added support for 0x0 matrices in https://github.com/pytorch/pytorch/pull/45832/commits/b15fedda0078a4a544f3f2f44e75a46288dc0eba.

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+
+  Tensor result_tmp = at::linalg_cond(self, opt_ord);
+  at::native::resize_output(result, result_tmp.sizes());
+  result.copy_(result_tmp);
```
is it possible to refactor this to avoid the copy? have `linalg_cond_helper` modify result in-place?

IvanYashchuk replied:
Yes, it's possible, but is it worth it? The performance and memory cost of this function are dominated by SVD and computing the inverse matrix.

mruberry replied:
We should do this but it doesn't need to be in this PR.

ngimel commented on test/test_linalg.py:
```diff
@@ -637,6 +637,121 @@ def run_test_case(input, p, dim, keepdim):
+                run_test_case(input, ord)
+
+        # test non-square input
+        input_sizes = [(16, 32), (32, 16), (2, 3, 5, 3), (2, 3, 3, 5)]
```
please add tests for empty batch size

IvanYashchuk replied:
Done.

ngimel commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,86 @@
+
+This function supports ``float``, ``double``, and only on CPU, ``cfloat`` and ``cdouble`` dtype for :attr:`input`.
+
+.. note:: For `p = {None, 2, -2}` :attr:`input` can be non-square. For other norm types :attr:`input` must be
```
this contradicts the above definition of "matrix norm of input times the matrix norm of the inverse of input", as non-square matrices don't have inverse.

IvanYashchuk replied:
Should I add that for `p = {None, 2, -2}` pseudoinverse is used instead of inverse?

The condition number based on 2-norm is κ₂  = || A⁻¹ ||₂ ||A||₂. For j singular values, of A, that are ordered from largest to smallest with 1...j, ||A||₂ = σ₁ and || A⁻¹ ||₂ = 1 / σⱼ. Incidentally, we can compute singular values for non-square matrices, but the definition of the condition number stays the same.

mruberry replied:
Yes, this is a helpful clarification (that I missed, thanks @ngimel!).

IvanYashchuk replied:
I updated the note in 5ed8a77.
```
.. note:: For `p = {None, 2, -2}` there is a relation between singular value decomposition and matrix norm.
          For this case :func:`torch.linalg.svd` is used for computing the condition number as the ratio of
          the largest and smallest singular values. Since :func:`torch.linalg.svd` is used :attr:`input` can be non-square.
          For other norm types :attr:`input` must be a square matrix or a batch of square matrices.
          If :attr:`input` does not satisfy this requirement then a RuntimeError will be thrown.
```

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1606,125 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // ord == ±1 ord == ±inf
+  // since at::inverse is used in the implementation, self has to be a tensor consisting of square matrices
+  // the same check as squareCheckInputs(self) but with a slightly more informative error message
+  TORCH_CHECK(self.size(-1) == self.size(-2),
```
Does it make sense to check that `ord` is allowable value here, otherwise the error will come from matrix norm and will be confusing?

IvanYashchuk replied:
I can add the checks here, but the documentation clearly says that the condition number is computed using matrix norms and error messages like below don't seem confusing to me.
```python
In [3]: torch.linalg.cond(a, p=3)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-1812a4cf0c1e> in <module>
----> 1 torch.linalg.cond(a, p=3)

RuntimeError: Order 3 not supported for matrix norm

In [4]: torch.linalg.cond(a, p='something')
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-1af0bc1a4547> in <module>
----> 1 torch.linalg.cond(a, p='something')

RuntimeError: Invalid norm order: something
```

mruberry replied:
It's not the end of the world, but since we're here already we may as well improve it.

IvanYashchuk replied:
Added checks in e407098.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk committed:
Fix typo norm -> cond

IvanYashchuk committed:
Rename ord -> p

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Remove 2-norm mentions in the docs, leaving just ratio of singular values

IvanYashchuk committed:
Use IntArrayRef for constructing result_shape

IvanYashchuk committed:
Allow batched input with zero dimensions; singular values are sorted so no need to use max and min

IvanYashchuk committed:
Added empty batch size test inputs

IvanYashchuk committed:
Revert changes to common_methods_invocations.py

IvanYashchuk committed:
Return 0 for 0x0 matrices

IvanYashchuk committed:
Added tests case for 0x0 matrices

Added complex dtype tests cases;
Renamed ord -> p

IvanYashchuk committed:
Updated note for p = 2, -2 to include reference to svd

IvanYashchuk committed:
Added checks for norm type

IvanYashchuk commented:
@mruberry, I updated this pull request:
- Modified the implementation to return zero for 0x0 matrices and added tests for that.
- Added checks whether the provided norm type `p` is valid.
- Updated documentation to include a reference to svd for explaining why non-square inputs are allowed for `p={None, -2, 2}`

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -339,6 +339,88 @@
+Computes the condition number of a matrix :attr:`input`,
+or of each matrix in a batched :attr:`input`, using the matrix norm defined by :attr:`p`.
+The condition number is defined as the matrix norm of
+:attr:`input` times the matrix norm of the inverse of :attr:`input`.
```
Should this be "inverse (or pseudo-inverse)" following the latest update?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -339,6 +339,88 @@
+
+This function supports ``float``, ``double``, and only on CPU, ``cfloat`` and ``cdouble`` dtypes for :attr:`input`.
+
+.. note:: For `p = {None, 2, -2}` there is a relation between singular value decomposition and matrix norm.
```
Suggest simplifying this to: 

"For \`p = {None, 2, -2}\` the condition number is computed as the ratio between the largest and smallest singular values computed using  :func:\`torch.linalg.svd\`. For these norms :attr:\`input\` may be a non-square matrix or batch of non-square matrices. For other norms, however, :attr:\`input\` must be a square matrix or a batch of square matrices, and if this requirement is not satisfied a RuntimeError will be thrown."

IvanYashchuk replied:
Your variant is nicer to read, thanks!

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -339,6 +339,88 @@
+
+Computes the condition number of a matrix :attr:`input`,
+or of each matrix in a batched :attr:`input`, using the matrix norm defined by :attr:`p`.
+The condition number is defined as the matrix norm of
```
Should this be something like,

"For norms ... this is defined as the matrix norm of ... And for norms ... this is defined as the ratio ..."

per the latest update?

IvanYashchuk replied:
Yes, I've changed that.

mruberry commented:
Update looks pretty good, @IvanYashchuk, just a last couple doc tweaks.

IvanYashchuk committed:
Updated documentation

IvanYashchuk commented:
@mruberry, I updated the docs per your suggested tweaks.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -339,6 +339,88 @@
+          For other norms, however, :attr:`input` must be a square matrix or a batch of square matrices,
+          and if this requirement is not satisfied a RuntimeError will be thrown.
+
+.. note:: If :attr:`input` is a non-invertible matrix then a tensor containing infinity will be returned.
```
Darn, I just realized this note will need to be updated, too, since non-invertbility doesn't apply to both cases, and the next sentence is redundant with the previous note.

Sorry I missed that on the last pass, @IvanYashchuk.

IvanYashchuk replied:
I updated this note mentioning that it applies only for some `p`.
```
.. note:: For ``p = {'fro', 'nuc', inf, -inf, 1, -1}`` if :attr:`input` is a non-invertible matrix then
          a tensor containing infinity will be returned. If :attr:`input` is a batch of matrices and one
          or more of them is not invertible then a RuntimeError will be thrown.
```

IvanYashchuk committed:
Changed the note on non-invertible input

Specified that this note applies only for p = {'fro', 'nuc', inf, -inf, 1, -1}

mruberry commented:
Lint failure looks real, unfortunately:

```
{
    path: 'torch/linalg/__init__.py',
    start_line: 354,
    end_line: 354,
    start_column: 141,
    end_column: 141,
    annotation_level: 'failure',
    message: '[B950] line too long (140 > 120 characters)'
  }
```

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg_cond

IvanYashchuk commented:
@mruberry I'm sorry about that. Seems like I need to install pre-commit hooks to avoid these situations.
Now it's fixed.
